### ISSUE TITLE
getUserMedia was supposed to work only via HTTPS, but web-dictaphone runs okay over HTTP on Android?

### ISSUE THREAD

**Issue by mavavilj** [mdn/dom-examples#244]
Based on:

https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia

the getUserMedia() method should be available only in secure contexts (HTTPS).

However, I am deploying web-dictaphone on CivetWeb (https://github.com/civetweb/civetweb) locally and it seems to run fine over HTTP.

The browsers even say it's truly an insecure context.

**Comment by mavavilj** [comment]
And yes it does say that localhost is considered secure, however I thought it would still require HTTPS.

**Comment by mavavilj** [comment]
Proof:

https://www.youtube.com/shorts/LbbDJivS58c

**Comment by mavavilj** [comment]
Or it means like it says on:

https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

> Locally-delivered resources such as those with http://127.0.0.1 URLs, http://localhost and http://*.localhost URLs (e.g. http://dev.whatever.localhost/), and file:// URLs are also considered to have been delivered securely.

But then it's confusing that Firefox still marks it as insecure.

**Comment by guest271314** [comment]
> But then it's confusing that Firefox still marks it as insecure.

Firefox doesn't mark HTTP as insecure.

**Comment by mavavilj** [comment]
> > But then it's confusing that Firefox still marks it as insecure.
> 
> Firefox doesn't mark HTTP as insecure.

It does.

**Comment by guest271314** [comment]
Where? I just created an HTTP server with `node` and successfully use `getUserMedia()` on http: protocol on Firefox Nightly 123. I didn't observe any notifications http: was "insecure".

**Comment by bsmth** [comment]
Thanks both. I think the question is answered based on https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure on localhost. I'm going to close for now, but if you think that's a mistake, please let me know. Thank you :)

**Comment by mavavilj** [comment]
> Thanks both. I think the question is answered based on https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure on localhost. I'm going to close for now, but if you think that's a mistake, please let me know. Thank you :)

I do not have the documentation for this problem anymore, but assuming the description in the first post holds, then that Mozilla link does not clarify the situation.

In the first post, I said: "The browsers even say it's truly an insecure context.". If the context would be secure, then a browser such as Firefox for Android would say it is a secure context (such as HTTP in localhost).

**Comment by bsmth** [comment]
Thanks for replying. I don't think there's a bug here, there may be some confusion here between what the browser shows in the UI on HTTP pages (on localhost) versus a secure context in terms of JS execution. You can try this, for example:

```js
if (window.isSecureContext) {
  console.log("Secure context");
} else {
  console.log("Not a secure context");
}
```

And check this for more information: https://developer.mozilla.org/en-US/docs/Web/API/Window/isSecureContext

Even on `http://localhost:5042/en-US/` I will get `true`

**Comment by mavavilj** [comment]
Okay, so why would the Firefox "lock symbol" (near the address bar) show a context still as insecure?

**Comment by bsmth** [comment]
> Okay, so why would the Firefox "lock symbol" (near the address bar) show a context still as insecure?
> 
> Or possibly, the context is secure, but there's a problem with the certificate (in which case it's actually still not secure)? Or it is then secure, but only because the certificate does not matter on localhost?

Yeah that's essentially the gist. The UI element in Firefox (the lock) is [about the TLS cert](https://support.mozilla.org/en-US/kb/secure-website-certificate). You can actually add one for localhost, but there isn't one by default, so you won't see a lock like on https sites unless you add your own cert. There's an interesting article here if you want some other reading material - https://web.dev/articles/when-to-use-local-https

In terms of "secure contexts", local resources are considered secure - they're already on your machine.

**Comment by mavavilj** [comment]
But why wouldn't Firefox show them as secure nevertheless?

**Comment by bsmth** [comment]
Localhost is considered a **secure context** even without TLS (on `http:`). See https://w3c.github.io/webappsec-secure-contexts/

**Comment by mavavilj** [comment]
So does the lock icon only reflect on the certificate and not secure contexts per se?
